From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mostly_Harmless <geek520170@gmail.com>
Date: Sat, 10 Oct 2026 16:30:21 +0800
Subject: [PATCH 6/7] fix(string): do not encode UTF-16 surrogates as UTF-8

A danmaku containing "&#55357;" (half of an emoji's UTF-16 pair, as
some clients send) was written as the bytes ED A0 BD, which is not valid
UTF-8, so strict readers reject the whole subtitle file. Treat U+D800 to
U+DFFF like other invalid code points and emit nothing.
---
 src/String/DanmakuFactoryString.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/src/String/DanmakuFactoryString.c b/src/String/DanmakuFactoryString.c
index d094743..89bcbaf 100644
--- a/src/String/DanmakuFactoryString.c
+++ b/src/String/DanmakuFactoryString.c
@@ -891,14 +891,14 @@ void unicode_to_utf8(unsigned int codepoint, char *utf8_buffer)
         utf8_buffer[1] = (char)(0x80 | (codepoint & 0x3F));
         utf8_buffer[2] = '\0';
     }
-    else if (codepoint <= 0xFFFF)
-    {
+    else if (codepoint <= 0xFFFF && (codepoint < 0xD800 || codepoint > 0xDFFF))
+    { /* U+D800-U+DFFF 是 UTF-16 代理项，不是字符，编码后不是合法的 UTF-8 */
         utf8_buffer[0] = (char)(0xE0 | (codepoint >> 12));
         utf8_buffer[1] = (char)(0x80 | ((codepoint >> 6) & 0x3F));
         utf8_buffer[2] = (char)(0x80 | (codepoint & 0x3F));
         utf8_buffer[3] = '\0';
     }
-    else if (codepoint <= 0x10FFFF)
+    else if (codepoint > 0xFFFF && codepoint <= 0x10FFFF)
     {
         utf8_buffer[0] = (char)(0xF0 | (codepoint >> 18));
         utf8_buffer[1] = (char)(0x80 | ((codepoint >> 12) & 0x3F));
